compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorLowering.cpp

```diff
@@ -183,6 +184,288 @@ struct SetMulAddFMF final : OpRewritePattern<vector::MultiDimReductionOp> {
+    // vector.mask.
+    if (op.isMasked() || op.getKind() != vector::CombiningKind::ADD) {
+      return failure();
+    }
+
+    VectorType lhsVecType = op.getLhsType();
+    VectorType rhsVecType = op.getRhsType();
+    if (lhsVecType.isScalable() || rhsVecType.isScalable()) {
+      return failure();
+    }
+
+    auto accVecType = dyn_cast<VectorType>(op.getAccType());
+    if (accVecType && accVecType.isScalable()) {
+      return failure();
+    }
+
+    if (!isa<FloatType>(lhsVecType.getElementType())) {
+      return failure();
+    }
+
+    Type accElemType = getElementTypeOrSelf(op.getAccType());
+
+    Location loc = op.getLoc();
+    Value lhs = op.getLhs();
+    Value rhs = op.getRhs();
+
+    if (lhsVecType.getElementType() != accElemType) {
+      Type promotedType = lhsVecType.clone(accElemType);
+      lhs = arith::ExtFOp::create(rewriter, loc, promotedType, lhs);
+      lhsVecType = cast<VectorType>(lhs.getType());
+    }
+
+    if (rhsVecType.getElementType() != accElemType) {
+      Type promotedType = rhsVecType.clone(accElemType);
+      rhs = arith::ExtFOp::create(rewriter, loc, promotedType, rhs);
+      rhsVecType = cast<VectorType>(rhs.getType());
+    }
+
+    Type elemTy = accElemType;
+    SmallVector<int64_t> redDims, parDims;
+    getReductionAndParallelLoopDims(op.getIteratorTypes(), redDims, parDims);
+    if (redDims.empty()) {
+      return failure();
+    }
+
+    // New indices: [reduction..., parallel...].
+    SmallVector<int64_t> indices;
+    indices.reserve(redDims.size() + parDims.size());
+    llvm::append_range(indices, redDims);
+    llvm::append_range(indices, parDims);
+
+    ArrayRef<int64_t> lhsShape = lhsVecType.getShape();
+    ArrayRef<int64_t> rhsShape = rhsVecType.getShape();
+    SmallVector<AffineMap, 4> maps = op.getIndexingMapsArray();
+    AffineMap lhsMap = maps[0];
+    AffineMap rhsMap = maps[1];
+    AffineMap accMap = maps[2];
+
+    // Broadcast operands for missing parallel dimensions.
+    unsigned numParallelDims = accMap.getNumResults();
+    SmallVector<int64_t> lhsReductionDims =
+        getReductionIndex(lhsMap, op.getIteratorTypes());
+    SmallVector<int64_t> rhsReductionDims =
+        getReductionIndex(rhsMap, op.getIteratorTypes());
+
+    unsigned numLhsDimToBroadcast =
+        numParallelDims - (lhsMap.getNumResults() - lhsReductionDims.size());
+    unsigned numRhsDimToBroadcast =
+        numParallelDims - (rhsMap.getNumResults() - rhsReductionDims.size());
+
+    SmallVector<int64_t> lhsBroadcastDims;
+    SmallVector<int64_t> lhsTranspose;
+    SmallVector<int64_t> rhsBroadcastDims;
+    SmallVector<int64_t> rhsTranspose;
+
+    for (int64_t dim : lhsReductionDims) {
+      lhsTranspose.push_back(numLhsDimToBroadcast + dim);
+    }
+    for (int64_t dim : rhsReductionDims) {
+      rhsTranspose.push_back(numRhsDimToBroadcast + dim);
+    }
+
+    for (unsigned i = 0; i < numParallelDims; i++) {
```

Comment:
```suggestion
    for (unsigned i = 0; i < numParallelDims; ++i) {
```
See https://llvm.org/docs/CodingStandards.html#prefer-preincrement